feat: Letterhead Scripts & fix disappearing header in pdf

marination opened the pull request:
`no-docs`
Port of https://github.com/frappe/frappe/pull/23263

**Setup:**
- Add the following Letterhead HTML
  ```html
  <div id="header-html">
	    <div class="header-content">
	        Page <span class="page"></span> of <span class="topage"></span>
	    </div>
  </div>
   ```
- Add the following header script in the letterhead. The idea is to have the header be in red in a pdf and in blue via print to make sure the script is visibly applied
  ```js
	var vars = {};
	var query_strings_from_url = document.location.search.substring(1).split('&');
	for (var query_string in query_strings_from_url) {
		if (query_strings_from_url.hasOwnProperty(query_string)) {
			var temp_var = query_strings_from_url[query_string].split('=', 2);
			vars[temp_var[0]] = decodeURI(temp_var[1]);
		}
	}
	var el = document.getElementsByClassName("header-content");
	
	if (vars["page"] == "1") {
		// "page" will be available only in pdf via wkhtmltopdf
		el[0].style.color = "red";
	} else {
	    el[0].style.color = "blue";
	}
  ```

**Result (Note the header):**
- PDF
   <img width="600" alt="Screenshot 2023-12-21 at 5 31 41 PM" src="https://github.com/frappe/frappe/assets/25857446/3c9702d7-3f44-4046-9bab-42127ba8e840">
- Print (does not have page numbers substituted in custom header because it is not renedered via wkhtmltopdf, treat this as an example)
  <img width="472" alt="Screenshot 2023-12-21 at 5 32 22 PM" src="https://github.com/frappe/frappe/assets/25857446/bb3bc6f2-f244-4f88-b090-b02566ee6f83">

marination committed:
fix: Disappearing letterhead header in pdf

- fix: Make sure 'header-html' is extracted from document after it is rendered as html and loaded in wkhtmltopdf options, else it goes missing

marination committed:
feat: Letterhead scripts

- feat: Allow script injection into header/footer.html to allow manipulation of styles using page numbers/args received by wkhtmltopdf
- misc: also validate letterhead scripts
- Include scripts in printview as well as pdf
- Add helper instructions & re-arrange fields

marination committed:
fix: Translate Letterhead information string

marination committed:
fix: Custom Script runs twice

- `frappe.get_print` calls `get_response_content("printview")` which calls `www/printview` that adds header and footer scripts
- Now it gets added again via `pdf_header_footer.html` which causes the script to run twice
- Header and Footer html are also added twice but then removed and handled in pdf.py (because printview + wkhtml processing)
- Misc: Use walrus operator
- Remove passed `letterhead` to insert script via pdf.py, printview handles it

maharshivpatel commented on frappe/utils/pdf.py:
Everything LGTM, i think instead of removing instances after render we can also just extract it in the first place.
```python
if content := soup.find(id=html_id).extract():
	# `header/footer-html` are extracted, rendered as html
	# and passed in wkhtmltopdf options (as '--header/footer-html')
	# Remove instances of them from main content after render_template
	for tag in soup.find_all(id=html_id):
		tag.extract()
```

maharshivpatel replied:
@marination, please add your thoughts. we can merge this either way.

marination replied:
@maharshivpatel That works I'll push this. Can confirm things work as expected

marination committed:
fix: Extract header/footer html into `content` to simplify things

- Extract header/footer html into `content` first, then remove any instances from the main content since it is already rendered via the extracted header/footer html

marination commented:
@maharshivpatel done

marination committed:
fix: Check if header/footer html exists before trying to extract

- issue: AttributeError: 'NoneType' object has no attribute 'extract', when no letterhead html was present
